[WIP] Add new BuildConfig form

jerolimov opened the pull request:
**Fixes**: 
Epic: https://issues.redhat.com/browse/ODC-5008
Stories:
https://issues.redhat.com/browse/ODC-6084 Create a form for editing BuildConfig resource
https://issues.redhat.com/browse/ODC-6085 Switch between YAML and Form view for editing BuildConfig

**Analysis / Root cause**: 
todo

**Solution Description**: 
todo

**Screen shots / Gifs for design review**: 
todo

**Unit test coverage report**: 
todo

**Test setup:**
todo

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

jerolimov commented:
/uncc kyoto
/cc @invincibleJai

divyanshiGupta commented on frontend/packages/dev-console/src/components/buildconfig/BuildConfigFormEditor.tsx:
```diff
@@ -0,0 +1,93 @@
+      <ImagesSection />
+      <EnvironmentVariablesSection namespace={namespace} />
+
+      <FormSection
+        title={t('devconsole~Advanced options')}
+        dataTest="section advanced-options"
+        fullWidth
+      >
+        <ProgressiveList
+          text={t('devconsole~Click on the names to access advanced options for')}
+          visibleItems={visibleItems}
+          onVisibleItemChange={handleVisibleItemChange}
+        >
+          <ProgressiveListItem name={t('devconsole~Triggers')}>
+            <TriggersSection namespace={namespace} />
+          </ProgressiveListItem>
+          <ProgressiveListItem name={t('devconsole~Secrets')}>
+            <SecretsSection namespace={namespace} />
+          </ProgressiveListItem>
+          <ProgressiveListItem name={t('devconsole~Run Policy')}>
+            <PolicySection />
+          </ProgressiveListItem>
+          <ProgressiveListItem name={t('devconsole~Hooks')}>
+            <HooksSection />
+          </ProgressiveListItem>
+        </ProgressiveList>
+      </FormSection>
```
Create a separate component for `Advanced options`

jerolimov replied:
Thanks for that idea, but I like that all "sections" are listed in one file and that this gives a code-reader an instant overview which sections, which are part of the advanced options and also which order to we use for the directly visible sections and the advanced options. I would like to keep this as it is.

rohitkrai03 commented on frontend/packages/console-dynamic-plugin-sdk/src/api/api-types.ts:
```diff
@@ -33,7 +33,7 @@ export type WatchK8sResources<R extends ResourcesObject> = {
 
 export type UseK8sWatchResource = <R extends K8sResourceCommon | K8sResourceCommon[]>(
-  initResource: WatchK8sResource,
+  initResource: WatchK8sResource | null,
```
Just curious, why is this needed?

jerolimov replied:
In BuildConfigFormPage I used this code to fetch and watch the BuildConfig only if a name is defined.

```tsx
  const isNew = !name || name === '~new';
  const [watchedBuildConfig, loaded, loadError] = useK8sWatchResource<BuildConfig>(
    isNew
      ? null
      : {
          kind: BuildConfigModel.kind,
          name,
          namespace,
        },
  );
```

`useK8sWatchResource` already supported to skip fetching if the parameter is null but the type definition doesn't allow this.

rohitkrai03 commented on frontend/packages/dev-console/src/components/buildconfig/__tests__/user-event.ts:
```diff
@@ -0,0 +1,68 @@
+// TODO: Replace this workaround with a working version of @testing-library/user-event
+// To check the order of events see also
+// https://codesandbox.io/s/user-event-playground-eo909?file=/src/index.js
+import { fireEvent } from '@testing-library/react';
```
Any reason you implemented your own version of `userEvent`?

jerolimov replied:
Yes, the user-event library crashs when calling type / click on our @console/shared or PatternFly fields. I created the issue https://issues.redhat.com/browse/ODC-6264 to track this and hope that we can remove this soon.

jerolimov replied:
I believe both, our code, maybe PF, but we can maybe also open a ticket at userEvent. We should do that after deeper investigation when working on the new ODC ticket.

divyanshiGupta commented on frontend/packages/dev-console/src/components/buildconfig/sections/ImagesSection.tsx:
```diff
@@ -0,0 +1,104 @@
+  );
+};
+
+export default ImagesSection;
```
Make separate files for each component.

jerolimov replied:
Is there really a benefit when extracting this into another file? `ImagesSection` and `ImageOption` are currently closely related and we have "supporting sub" components in other code areas. But I can do this if this will block someone from giving an lgtm.

rohitkrai03 replied:
I think it's okay to co-locate components that are tightly coupled in some cases.

rohitkrai03 commented on frontend/packages/dev-console/src/components/buildconfig/sections/EditorField.tsx:
```diff
@@ -0,0 +1,49 @@
+      validated={error ? 'error' : 'default'}
+      isRequired={required}
+    >
+      <MonacoEditor
```
Would be good to lazy load this component to avoid pulling whole of monaco editor as dependency into main vendor bundle.

jerolimov replied:
Does this change the vendor bundle? I need to check how I can test this.

The complete route is currently lazy loaded via `packages/dev-console/src/plugin.tsx`, @invincibleJai noticed below that we should switch to dynamic plugin instead.

I can check this or maybe we can do both as an follow up ticket?

rohitkrai03 replied:
If the complete route is lazy loaded it should be fine. But if this was used as a static plugin and imported through plugin.tsx it would create issues.

rohitkrai03 replied:
Maybe we can even move this to formik form fields later.

invincibleJai commented on frontend/packages/dev-console/src/components/buildconfig/BuildConfigForm.tsx:
```diff
@@ -0,0 +1,117 @@
+//   convertEditFormToDeployment,
+// } from './utils/edit-deployment-utils';
+
+const EditDeploymentForm: React.FC<FormikProps<BuildConfigFormikValues> & {
```
file name is `BuildConfigForm.tsx` and component is `EditDeploymentForm`

invincibleJai commented on frontend/packages/dev-console/src/components/buildconfig/BuildConfigForm.tsx:
```diff
@@ -0,0 +1,117 @@
+} from './form-utils/form-utils';
+import { BuildConfigFormikValues } from './form-utils/types';
+import { BuildConfig, BuildConfigModel } from './types';
+// import {
+//   convertDeploymentToEditForm,
+//   convertEditFormToDeployment,
+// } from './utils/edit-deployment-utils';
```
commented code

invincibleJai commented on frontend/packages/dev-console/src/components/buildconfig/BuildConfigForm.tsx:
```diff
@@ -0,0 +1,117 @@
+
+  const { editorType, yamlData } = values;
+
+  const isStale = false; // oldBuildConfig.metadata.resourceVersion !== formData.resourceVersion;
```
looks like this needs update

invincibleJai commented on frontend/packages/dev-console/src/components/buildconfig/__tests__/BuildConfigFormPage.spec.tsx:
```diff
@@ -0,0 +1,61 @@
+      match: { params: { ns: 'a-namespace', name: '~new' } },
+    } as BuildConfigFormPageProps;
+
+    render(
+      <Wrapper>
+        <BuildConfigFormPage {...props} />
+      </Wrapper>,
+    );
+  });
```
there are no assertions here and below it blocks

jerolimov replied:
It was placeholders, but I updated them now with some different cases.

invincibleJai commented on frontend/packages/dev-console/src/components/buildconfig/form-utils/form-utils.ts:
```diff
@@ -0,0 +1,36 @@
+import { BuildConfigFormikValues } from './types';
+
+export const convertBuildConfigToFormData = (buildConfig: BuildConfig): BuildConfigFormikValues => {
+  // eslint-disable-next-line no-console
+  console.log('xxx convertBuildConfigToFormData buildConfig in', buildConfig);
```
nit: console logs clean up

invincibleJai commented on frontend/packages/dev-console/src/components/buildconfig/sections/SecretsSection.tsx:
```diff
@@ -0,0 +1,57 @@
+
+export type SecretsSectionFormData = {
+  formData: {
+    secrets: {
+      secrets: {
```
nested `secrets` in form data ?

jerolimov replied:
Removed nested :+1:

invincibleJai commented on frontend/packages/dev-console/src/plugin.tsx:
```diff
@@ -208,13 +208,25 @@ const plugin: Plugin<ConsumedExtensions> = [
+    },
+  },
+  {
+    type: 'Page/Route',
+    properties: {
+      exact: true,
+      path: ['/k8s/ns/:ns/buildconfigs/~new/form', '/k8s/ns/:ns/buildconfigs/:name/form'],
+      loader: async () =>
+        (
+          await import(
+            './components/buildconfig/BuildConfigFormPage' /* webpackChunkName: "dev-console-buildconfig-form" */
```
can move this as dynamic extension

divyanshiGupta commented:
Please remove the commented code from the PR.

debsmita1 commented:
The tag is not persisted 
![buildconfig-edit](https://user-images.githubusercontent.com/22490998/130443889-6fce0ee1-c59c-4a19-9d83-c52cea48ae06.gif)

invincibleJai commented:
@jerolimov I don't see data being persisted between form/yaml i.e for `Build from`

![Kapture 2021-08-23 at 18 55 59](https://user-images.githubusercontent.com/5129024/130455592-1a47c652-9474-44b3-bdc1-f357dc489d6e.gif)

debsmita1 commented:
The BuildConfig yaml data shows that it's of docker strategy type but in the edit mode it shows source strategy type

![strategy](https://user-images.githubusercontent.com/22490998/130458017-23edce97-c542-4e06-ada2-31742faa5b18.gif)

invincibleJai commented on frontend/packages/dev-console/src/components/buildconfig/BuildConfigForm.tsx:
```diff
@@ -0,0 +1,115 @@
+  const onReload = React.useCallback(() => {
+    setStatus({ submitSuccess: '', submitError: '' });
+    setErrors({});
+    if (editorType === EditorType.YAML) {
+      // TODO setFieldValue('formData.resourceVersion', oldBuildConfig.metadata.resourceVersion);
+      setFieldValue('yamlData', safeJSToYAML(oldBuildConfig, 'yamlData', { skipInvalid: true }));
+    } else {
+      // setFieldValue('formData', convertDeploymentToEditForm(resource));
+    }
```
@jerolimov this still needs change?

jerolimov replied:
Updated this part :+1:

divyanshiGupta commented:
@jerolimov some of my above comments are not yet resolved. PTAL.

debsmita1 commented:
cannot create secret from the advanced git options
![secret](https://user-images.githubusercontent.com/22490998/130577320-6290caf0-2aa4-48c7-9a51-13ab93ff8bfe.gif)

jerolimov commented:
/test frontend

jerolimov commented:
> ✖  addFlow/developer-catalog-details.feature

/retest

jerolimov commented:
Since the epic acceptance criteria of [ODC-5008](https://issues.redhat.com/browse/ODC-5008) have not changed and the epic has a doc-ack and px-ack, we approve this on behalf of doc and PX team.

/label docs-approved
/label px-approved

rohitkrai03 commented on frontend/packages/dev-console/src/components/buildconfig/form-utils/__tests__/convert-to-buildconfig.spec.ts:
```diff
@@ -0,0 +1,663 @@
+          dockerfilePath: 'Dockerfile',
+        },
+        configMaps: [{ configMap: { name: 'a-configmap' }, destinationDir: '/configmap' }],
+        // secrets: [{ secret: { name: 'a-secret' }, destinationDir: '/secret' }],
```
Any reason to comment this out?

jerolimov replied:
Nope. Removed it.

rohitkrai03 commented on frontend/packages/dev-console/src/components/buildconfig/form-utils/initial-data.ts:
```diff
@@ -0,0 +1,144 @@
+            // tag: '',
+            tagObj: '',
+            // ports: [],
+            // isRecommending: false,
```
Remove commented code if not needed.

rohitkrai03 commented on frontend/packages/dev-console/src/components/buildconfig/sections/HooksSection.tsx:
```diff
@@ -0,0 +1,95 @@
+          options={{
+            lineHeight,
+            readOnly: false,
+            // folding: false,
```
```suggestion
```

rohitkrai03 commented on frontend/packages/dev-console/src/components/buildconfig/sections/__tests__/EnvironmentVariablesSection.spec.tsx:
```diff
@@ -0,0 +1,171 @@
+      </Wrapper>,
+    );
+
+    renderResult.getByTestId('section environment-variables');
```
Any reason you're not using `screen.getByTestId` here?

jerolimov replied:
renderResult provides also debug and container access.

rohitkrai03 commented on frontend/packages/dev-console/src/components/buildconfig/sections/__tests__/ImagesSection.spec.tsx:
```diff
@@ -0,0 +1,400 @@
+import ImagesSection, { ImagesSectionFormData } from '../ImagesSection';
+
+/*
+jest.mock('@console/internal/components/utils/firehose', () => ({
+  Firehose: ({ resources, children, ...otherProps }) => {
+    return require.requireActual('react').cloneElement(children, {
+      ...otherProps,
+      reduxIDs: ['project.openshift.io~v1~Project'],
+      resources: {
+        project: {
+          kind: 'Project',
+          loaded: true,
+          loadError: '',
+          data: [
+            { metadata: { name: 'project-a' } },
+          ],
+        },
+      }
+    });
+  },
+}))
+*/
```
Why is this commented out?

jerolimov replied:
Doesn't worked yet, it is required for the test which is marked with .skip at the moment. I'm currently try to fix this test.

jerolimov replied:
Solved this now and clean-up-ed all tests.

jerolimov commented on frontend/packages/dev-console/src/components/buildconfig/BuildConfigFormPage.tsx:
```diff
@@ -0,0 +1,55 @@
+      }
+    : watchedBuildConfig;
+
+  const title = isNew ? t('devconsole~Create BuildConfig') : t('devconsole~Edit BuildConfig');
```
It is not linked, but it works "not bad" so I would recommend keep that condition so that we can suggest to add this in the future.

jerolimov commented on frontend/packages/dev-console/src/components/buildconfig/EditBuildConfig.tsx:
```diff
@@ -0,0 +1,114 @@
+      }
+    } catch (error) {
+      const err = error.message;
+      helpers.setStatus({ submitSuccess: '', submitError: t('devconsole~{{err}}', { err }) });
```
Thought that as well, changed to:

```tsx
helpers.setStatus({ submitSuccess: '', submitError: err.message });
```

OK?

debsmita1 commented:
the `mount point` value is not persisted, also is not used while updating the resource

![mountpoint](https://user-images.githubusercontent.com/22490998/130664966-35fbdbf1-69d3-4684-8c85-5517ab768a19.gif)

unable to use the samples
![samplesbc](https://user-images.githubusercontent.com/22490998/130665355-1911e23f-5af5-44e2-baac-fe9e742421be.gif)

jerolimov commented:
e2e issues?
/hold

jerolimov commented:
> "The tag is not persisted" by @debsmita1 

solved

> General persisted between form/yaml issue by @invincibleJai 

solved

> The BuildConfig yaml data shows that it's of docker strategy type but in the edit mode it shows source strategy type by @debsmita1 

The form in OpenShift 3.x doesn't support an option to change the strategy directly. Source type and strategy type are two things. But they are aligned to each other.

Because we support only editing of BuildConfigs at the moment, I think its okay because the code should not change this types. If we want support "Create BuildConfig" in the future (my code tries to prepare this as best as possible) we should maybe also show a Dropdown for the Strategy type. This mapping to the resource spec makes it also easier to understand.

> cannot create secret from the advanced git options by @debsmita1 

solved

> the mount point value is not persisted, also is not used while updating the resource by @debsmita1 

solved

karthikjeeyar commented:
On saving the form, `Git Type` field appears below the git repo url, Usually In Add flow Git Type field appears only when the unknown source control is detected. 

![buildconfig-gitType](https://user-images.githubusercontent.com/9964343/130727632-a52eec44-b4f0-4b30-9235-946dd3051685.gif)

karthikjeeyar commented:
Trigger Type field is not persisted.
![tirgger_type_not_persisted](https://user-images.githubusercontent.com/9964343/130727949-668a5cfa-6fbb-42f1-92e0-2ec6e2e70bdc.gif)

karthikjeeyar commented:
Choosing `None`  in build From field seems to be not working. The backend validation for sourceStrategy is complaining from.name is required field.

![None_not_working](https://user-images.githubusercontent.com/9964343/130728005-7a657541-2953-4d04-ad2c-b79276265e23.gif)

jerolimov commented:
Thanks @karthikjeeyar for this findings :+1: 

> On saving the form, Git Type field appears below the git repo url, Usually In Add flow Git Type field appears only when the unknown source control is detected.

It was shown then the type field was touched which also happen when the form is submitted. I changed this now in the shared GitSection component.

> Trigger Type field is not persisted

It was not shown correctly after loading the data. The types doesn't match the dropdown options. This is also fixed.

> Choosing None in build From field seems to be not working. The backend validation for sourceStrategy is complaining from.name is required field.

Thanks again. The form doesn't show the None option anymore when the spec.strategy.type is "Docker" similar to 3.x.

invincibleJai commented:
@jerolimov hey I observed this, did git import and then edit build config

![image](https://user-images.githubusercontent.com/5129024/130804996-d09c9178-a88c-46ca-984c-b145c439d271.png)

invincibleJai commented:
/lgtm

verified the changes, works as expected 

cc @rohitkrai03

jerolimov commented:
> e2e issues?
> /hold

hold was from myself after e2e issues yesterday.

/unhold

rohitkrai03 commented:
/assign

jerolimov commented:
> Overall works well but found one issue with monaco editor in hooks section. It seems like it's not loading up correctly.

Thanks @rohitkrai03, it's fixed.

`EditorField` now sets also the "console" theme and initialize it. As discussed, I also added a debounce to the `onChange` handler. [See latest diff](https://github.com/openshift/console/compare/de33482595c17a1b76799dd19e36cb5c04c4b99f..ec16d34ce5374d5357af6c24d6b10d6ea29f6808)
